python/tank/platform/software_launcher.py

```diff
@@ -205,72 +235,170 @@ def display_name(self):
+        Returned as a string, for example `2015` or `2015.3.sp3`.
         Returns ``None`` if no constraint has been set.
-        Also see related method :meth:`~is_version_supported`.
         """
         # returns none by default, subclassed by implementing classes
         return None
 
-    ##########################################################################################
-    # abstract methods
+    @property
+    def products(self):
+        """
+        A list of product names limiting executable discovery.
+
+        Example::
 
-    def scan_software(self, versions=None):
+            ["Houdini", "Houdini Core", "Houdini FX"]
         """
-        Performs a scan for software installations.
+        return self._products
 
-        :param list versions: List of strings representing versions
-                              to search for. If set to None or [], search
-                              for all versions. A version string is
-                              DCC-specific but could be something
-                              like "2017", "6.3v7" or "1.2.3.52"
-        :returns: List of :class:`SoftwareVersion` instances
+    @property
+    def versions(self):
         """
-        raise NotImplementedError
+        A list of versions limiting executable discovery.
+
+        Example::
+
+            ["15.5.324", "16.0.1.322"]
+        """
+        return self._versions
+
+    ##########################################################################################
+    # abstract methods
 
     def prepare_launch(self, exec_path, args, file_to_open=None):
         """
         Prepares the given software for launch
 
         :param str exec_path: Path to DCC executable to launch
         :param str args: Command line arguments as strings
-        :param str file_to_open: (optional) Full path name of a file to open on launch
+        :param str file_to_open: (optional) Full path name of a file to open on
+            launch
         :returns: :class:`LaunchInformation` instance
         """
         raise NotImplementedError
 
+    def _glob_and_match(self, match_template, tokens_expressions):
+        """
+        This is a helper method that can be invoked in an implementation of :meth:`scan_software`.
+
+        The ``match_template`` argument provides a template to use both for globbing files and then pattern
+        matching them using regular expressions provided by the ``tokens_expressions`` dictionary.
+
+        The method will first substitute every token surrounded by `{}` from the template with a ``*``
+        for globbing files. It will then replace the tokens in the template with the regular expressions
+        that were provided.
+
+        In the following example::
+
+            self._glob_and_match(
+                "//network/softwares/Nuke{full_version}/Nuke{major_minor_version}.exe",
+                {
+                    "full_version": r"([\d.v]+)",
```

Comment:
Now I'm wondering if we should imply the capture and keep these regex as simple as possible. 

```
self._glob_and_match(
    "//network/softwares/Nuke{full_version}/Nuke{major_minor_version}.exe",
    {
        "full_version": r"[\d.v]+",
        "major_minor_version": r"[\d.]+"
    }
 )
```

Even if the returned keys aren't used, it would keep the interface and return value simple.